clarity/src/vm/types/signatures.rs

```diff
@@ -528,28 +600,124 @@ impl TypeSignature {
                 }
             }
             TupleType(ref tuple_sig) => {
-                if let TupleType(ref other_tuple_sig) = other {
+                if let TupleType(ref other_tuple_sig) = &other {
                     tuple_sig.admits(other_tuple_sig)
                 } else {
                     false
                 }
             }
-            PrincipalType => {
-                if other == &PrincipalType {
-                    true
-                } else if let CallableType(CallableSubtype::Principal(_)) = other {
-                    true
+            CallableType(CallableSubtype::Principal(_)) => {
+                panic!("callable principal should never be asked to admit.")
+            }
+            ListUnionType(_) => panic!("ListUnionType should never be asked to admit."),
+            IntType | UIntType | BoolType | CallableType(CallableSubtype::Trait(_)) => {
+                &other == self
+            }
+            PrincipalType => {&other == self},
+            NoType => panic!("NoType should never be asked to admit."),
+            #[cfg(feature = "fuzzing")]
+            TypeSignature::TraitReferenceType(_) => {
+                panic!("TraitReferenceType should only be used for testing 2.05.")
+            }
+        }
+    }
+
+    #[cfg(feature = "fuzzing")]
+    pub fn old_admits_type(&self, other: &TypeSignature) -> Result<bool> {
+        let other = match other.concretize() {
```

Comment:
We shouldn't need to call `concretize` here because the types that require being concretized should not show up in the "old" types (`ListUnionType` and `CallableType`).